### PR TITLE
Add tensor blog post

### PR DESCRIPTION
**Pull request by lesters**
@kkraune Please review.

### CODE REVIEW

File: _posts/2021-09-30-computing-with-tensors.md [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+In Vespa we have introduced a tensor formalism that differs from what one
+usually finds in most modern machine learning frameworks today. The main
+differences are:
+
+- Named dimensions
+- Unified sparse and dense tensor types
+- A small but very powerful set of core functions
+
+In this blog post, we'll explore the differences and the benefits of tensors in
+Vespa. We will also introduce the recently released [tensor
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:31]: compared to what? maybe just simplify:
```suggestion
In this blog post, we'll explore different aspects of tensors in Vespa.
We will also introduce the [tensor
```

File: _posts/2021-09-30-computing-with-tensors.md [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+playground](https://docs.vespa.ai/playground/), a tool to get familiar with and
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:32]: just realized, we should have a Vespa logo on the playground site, too (unrelated to this post, though)

File: _posts/2021-09-30-computing-with-tensors.md [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+inference for all computation, catching errors at an early compile-time stage
+rather than during runtime.
+
+Later in the post, we'll see how this enables arbitrarily complex computation
+with only a minimal, concise set of core operations.
+
+
+## Sparse and dense tensors
+
+The tensor as a multi-dimensional array is often considered to be dense. This
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:98]:
```suggestion
The tensor as a multi-dimensional array is often considered to be _dense_. This
```

File: _posts/2021-09-30-computing-with-tensors.md [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+means that all combinations of dimension indexes have a value, even if that
+value is `0` or `NaN`. However, a tensor with many such values could be
+represented more efficiently as a sparse tensor, where only the non-empty values
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:101]:
```suggestion
represented more efficiently as a _sparse_ tensor, where only the non-empty values
```

File: _posts/2021-09-30-computing-with-tensors.md [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+are defined. This can lead to considerable savings in space.
+
+Unfortunately, the internal representation of sparse tensors in most frameworks
+makes them incompatible with regular dense tensors. This leads to an entirely
+separate set of functions operating on sparse and dense tensors, with functions
+to convert between the two.
+
+Vespa supports dense, sparse, and tensors that contain both dense and sparse
+dimensions, called a mixed tensor. A dense tensor is a tensor containing only
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:110]:
```suggestion
dimensions, called a _mixed_ tensor. A dense tensor is a tensor containing only
```

File: _posts/2021-09-30-computing-with-tensors.md [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+# The tensor formalism
+
+In this section, we'll describe Vespa's full tensor formalism, including
+tensors and tensor types, and the core set of tensor operations.
+
+## Tensor types
+
+A *tensor type* consists of a *value type* and a set of *dimensions*. The value type is a numeric data type: double, float, int8, bfloat16, etc. The default value type is double.
+
+A dimension consists of a *name* and a *type*. The type can be either *indexed* or *mapped*, and, if indexed, can optionally include a *size*. The number of dimensions in the tensor is called its *order*. A tensor without dimensions (0-order) is a single scalar value. A tensor with one dimension (first-order) is a vector. A tensor with two dimensions (second-order) is a matrix.
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:225]: I learned something new then - I thought indexed tensors always had a size for each dimension. maybe add an example with an indexed tensor without size for a dimension, too

File: _posts/2021-09-30-computing-with-tensors.md [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+from `B`, and so on. The lambda expression `f(a,b)(a * b)` defines these values
+should be multiplied. This is called the element-wise product.
+
+The actual inner (dot) product requires these values to be summed. Thus we can
+add a `sum` operation so the expression becomes `sum(A * B)`. The `sum`
+operation is actually a `reduce` operation, so this expands fully to the
+following:
+
+```
+reduce(join(A, B, f(a,b)(a,b)), sum)
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:348]:
```suggestion
reduce(join(A, B, f(a,b)(a * b)), sum)
```

File: _posts/2021-09-30-computing-with-tensors.md [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+In this post, we've presented Vespa's tensor language. This formalism provides
+generic computation over dense and sparse tensors, with full static type
+inference, using just eight foundational tensor functions. Named dimensions are
+used to add semantic information and achieve generality, and index labels
+provide support for models working with strings.
+
+In Vespa, this formalism makes numerical computation easier to express,
+understand, and optimize.
+
+For more information, take a look at Vespas [tensor
```
Comment (kkraune) [inline _posts/2021-09-30-computing-with-tensors.md:472]:
```suggestion
For more information, take a look at Vespa's [tensor
```

**Review by kkraune — APPROVED** [review]
💯

File: _posts/2021-09-30-computing-with-tensors.md (hunk 2 above)
Comment (lesters) [reply _posts/2021-09-30-computing-with-tensors.md:32]: Good idea!

File: _posts/2021-09-30-computing-with-tensors.md (hunk 6 above)
Comment (lesters) [reply _posts/2021-09-30-computing-with-tensors.md:225]: This is only really supported on the Java side at the moment, but important for stateless model evaluation. So I'll think I'll let this be as it is for now.

Comment (kkraune) [reply _posts/2021-09-30-computing-with-tensors.md:225]: all right!

**Pull request merged**